### DISCUSSION TITLE
[Proposal]: Allow Method Calls to follow Property Setters inside Init-Blocks

### DISCUSSION THREAD

**Discussion by najak3d** [dotnet/csharplang#5728] [Language Ideas]
There are plenty of examples where construction of an object involves more than just property settings.

This request is very simple, and tightly scoped, making it an easy addition to compiler logic, and should produce zero confusion for developers who would take advantage of this.

It is a *replacement* for Fluent-pattern, and can, in most cases, be used entirely in-lieu of Fluent pattern methods (which often require you to write many extension methods, or Builder classes to achieve the simplified notation.

In short, I am proposing that we allow any "void method" to be called from within the Init-Block, plus any methods prefixed by "init" (per proposal #3578).   The "init" methods must always precede the other method calls.

Here's the pattern I am requesting:

```c#
new MyType()
{
     InitProperty1 = value1,
     InitProperty2 = value2,
     .InitMethod1(args),
     .InitMethod2(args),
     .NonInitProperty1 = value3,
     .NonInitProperty2 = value4,  // <<== **READONLY FIELDS ARE FROZEN AFTER THIS LINE OF CODE
                                   // BECAUSE THE NEXT LINE CALLS A NON-INIT-METHOD.
     .NonInitMethod1(args),  // Can be extension methods here too.  All methods must return VOID
     .NonInitMethod2(args),
     .EventType1 += HandleEventType1,
     // Here are some  actual named examples of members that would be called here
     .Clicked += HandleClicked,
     .SetProperty("propName", propValue),   // this is for "attached properties" common in UI's (WPF, Xamarin, Avalonia)
     .ApplyStyleByName("styleName"),
     .SetGridLoc(column, row)  // wrapper for another attached property
     .Bind("propName", modelObject, Mode.TwoWay)
     .PrepareForUse() 
     .AddChildren(
             new ChildType()
             {
                    // property setters, followed by method calls
             },
             new ChildType2()
             {
                    // property setters, followed by method calls
             }
     )
}
```
Note that here I've included "event handlers" in this example, which fully accommodates the concerns being addressed by proposal here:
   [https://github.com/dotnet/csharplang/issues/5176](https://github.com/dotnet/csharplang/issues/5176)

This proposal is also to be combined with #3578, and tightens the spec, and forces all Init-Methods to be called from within the Init-Block, which makes the modification to compiler logic easier to implement, as it doesn't have to be concerned with what cannot be done in a less confined code context.

NOTE:  Unlike Fluent pattern, this notation does not require ANY extension methods or builders to enable the notation. 

Above, I am proposing that all method calls after the "readonly" only properties are frozen, will operate the EXACT SAME as it would if all ".method()" calls were simply called from outside the Initialization block using a local variable reference.   

**CON**: This proposal can alter the nature/expectation for the Init-Block itself, in that the methods would be executed AFTER the object was fully instantiated, rather than before.  It should be noted that the "Children {}" collection initializer already works in this fashion (i.e. operates on an fully constructed object).

**Restriction**:   Only methods that return **"void"** are permitted to be called, to avoid encouraging devs to ignore return values.   If the user really wants to call a non-void method, they can create an extension method that fills the gap for targeted non-void methods.

**Comment by jmarolf** [comment]
As per the rules laid out for this repo [here](https://github.com/dotnet/csharplang#c-language-design)

> For new feature proposals, however, please raise them for [discussion](https://github.com/dotnet/csharplang/issues?q=is%3Aopen+label%3ADiscussion+sort%3Aupdated-desc), and only submit a proposal as a pull request if invited to do so by a member of the Language Design Team (a "champion").

I am converting this to a discussion

**Comment by jmarolf** [comment]
> The proposal is for the above to operate the EXACT SAME as it would if all ".method()" calls were simply called from outside the Initialization block using a local variable reference.

@najak3d from this statement I am inferring that only instance methods are allowed to be called? Can I also call extension methods? What if the method does not return `void` or `MyType` but some other value `T`, is it ignored?

I would like a full example with either an existing type from the base-class-library or your own type where you have written out its full definition. The semantics of your proposal are not clear in its current form.

**Reply by najak3d** [reply]
There is no reason NOT to allow Extension methods, as at the point where methods are being called, the object has already been constructed.   My proposal is for this to be short-hand notation to work the same as the long-hand notation (as if the method were called on the constructed object outside of the Init-block).

If it returns a T type, it is lost.   That is fine.   If you need to do something with a returned type, then this notation will not do that for you (which is equivalent to Fluent-pattern).

This is merely intended to be a replacement for Fluent-pattern.   Achieves nearly the same succinct syntax, but without forcing you to create a bunch of extensions that return back the object for each call.   It makes "builder logic" a natural part of the C# language, and this change is confined to the Init-Block alone.

**Reply by jmarolf** [reply]
So if a method returns a `Task` there is no way to `await` that method?

**Reply by CyrusNajmabadi** [reply]
> but without forcing you to create a bunch of extensions

As per the other discussion, this was a choice on your part. There are much more trivial paths you could go find that would not have involved automating the creation of thousands of extension methods. I showed a simple four live extension method that would accomplish the same thing. 

So we'd have to weigh the high cost and complexity of this feature against, well, you writing 4 lines of code. That's what this proposal has to compete against.

**Comment by HaloFour** [comment]
This seems like it's just repeating the same thing over and over again and expecting a different response.

**Reply by najak3d** [reply]
My sample code already approximates code I use for Avalonia UI construction, but pulls in one method that I've used in the past for complex objects called "PrepareForUse()".

For these scenarios, instantiation is a two-step process, often.   First you might just deserialize something, and then later might call "PrepareForUse" which loads things into RAM needed for actual use (before you activate it).

In some cases, you'll do it at the same time as deserialization.  So I added "PrepareForUse()" to the above, although it does not appear in my Avalonia code.  Otherwise stuff like applying Styles, margins, and Binding -- this is how it looks inside Avalonia.

**Reply by najak3d** [reply]
> This seems like it's just repeating the same thing over and over again and expecting a different response.

To a point yes, but it's not about "expecting different results" -- it's about the impact of starting over with a clean-thread, and eliminate the meandering of the original discussion thread.   And there is value in doing that.

**Reply by jmarolf** [reply]
> My sample code already approximates code I use for Avalonia UI construction, but pulls in one method that I've used in the past for complex objects called "PrepareForUse()".

Please ensure your proposal has all the details necessary for someone to understand it. If you could at least link to the Avalonia types you are using that would be helpful. As it stands, I have no insight into what these types are, what these methods do, or where they come from if I am not intimately familiar with Avalonia.

**Reply by najak3d** [reply]
Before I do that -- I may have just rationalized why there is no good path forward.   See my response below.

**Reply by CyrusNajmabadi** [reply]
> and eliminate the meandering of the original discussion thread.

The discussion didn't meander. It was a thorough critique of the proposal with lots of still relevant points about this issues here, and a huge amount of information on why the *current* justifications for needing this we're coming up short.  :-)

**Comment by najak3d** [comment]
Regarding whether or not to have the ".method(...)" calls occur before or after construction is completed -- what are the dangers of calling methods before construction is completed?  (calling any accessible method)   Vs. having the method be called on the fully constructed object?

We could just treat those ".method()" calls as though they were called from inside of a constructor.   What is the inherent dangers to be concerned about?   The reason I'm asking, is that at least then, the coder can view everything inside the "{...}" init-block as "stuff done before construction is completed", to follow that mental paradigm.

So it would just the same as if you packed all that code directly into the top-level's class constructor.
```c#
new TopLevelClass() { Name = name, .PrepareForUse() }
```

Would be the same as:
```c#
public TopLevelClass(string name) { Name = name;  PrepareForUse(); } 
```

I'm entertaining this approach, due to many here being concerned with ensuring that the "meaning of an init-block" is clearly defined -- in that the object is not fully constructed until the init-block has completed.

So if we decided that it's best for everything inside the init-block to run the same as if it were hard-coded into the top-level class's constructor, then this proposal would accommodate that concern.

And so if there are issues with this modification to the proposal - then it seems more clear why this proposal seems to be getting little support, because:

1. If ".method()" are treated as "operating on the fully-constructed object", then this breaks the expectations of what an Init-Block is.
2. And if ".method()" operate on an object not-fully constructed, this might cause other risks/complications, making the effort "not a good idea".

I'm guessing that this is how many of the experienced folks here view this proposal.   #1 is bad idea.  #2 is bad idea.   Therefore, there is no reason to move forward with this proposal.

Am I correctly assessing the root cause of the rejection I'm seeing here?

**Reply by najak3d** [reply]
If there are issues with making .method() calls on an object before construction is complete, I'd like to see a few examples, to illustrate the tangible concerns regarding this approach.

**Reply by najak3d** [reply]
In short, I'm willing to accept defeat on the idea of having methods inside the init-block operate on the fully-constructed-object, because this notion breaks the "expectation" of what an Init-Block really is.

So - that limits this proposal to only having method operate on the object before construction is complete.  In which case I'd need to revise the proposal.   But if I can see how it'll be shot-down via example -- then I can just close this whole discussion, and not bring it up again.

**Reply by CyrusNajmabadi** [reply]
> Am I correctly assessing the root cause of the rejection I'm seeing here?

No. You are not. Please reread some of the stuff I wrote to you (including on discord and especially about -100). 

One of the points I've been trying to make clear is that a solution had to start with a clearly defined problem that people can agree on. In many cases you've claimed there is a problem here that is widespread and extremely detrimental, but you've only been able to point at your own team's issues with a particular automation strategy that you decided on. 

This is not a good defense for your position as it basically makes the argument: we did something super extreme, and we're unwilling to fix that first. Instead, we want a language change to make our extreme solution unnecessary. 

However, that argument fails to realize there are other truly trivial things you could have done already that wouldn't require the thousands and thousands of hours that go into a language change.  We have to compare your proposal against that alternative.

**Comment by CyrusNajmabadi** [comment]
```c#
 .Clicked += HandleClicked
     .SetProperty("propName", propValue),   // this is for "attached properties" common in UI's (WPF, Xamarin, Avalonia)
     .ApplyStyleByName("styleName"),
     .SetGridLoc(column, row)  // wrapper for another attached property
     .Bind("propName", modelObject, Mode.TwoWay)
     .PrepareForUse() 
     .AddChildren(
             new ChildType()
             {
                    // property setters, followed by method calls
             },
             new ChildType2()
             {
                    // property setters, followed by method calls
             }
     )
```

I don't get much of this proposal. 

`+=` is already handled by my proposal. 

`.SetProperty("prop", value)` can already be done today with `["p"] = v`. 

`.AddChildren(x, y)` should just be `Children = { x, y }`

There appears to be a lot of effort to reinvent a different way to do things that have been idiomatic in c# for 15+ years now.

**Comment by CyrusNajmabadi** [comment]
> if you are initializing children INSIDE the Init-Block (which is already supported by C#), then how do you call Methods upon the Parent object BEFORE adding children? 

Why are you doing this?  This really goes to my very first critique of your proposal:

> Yes. Methods imply state mutation as opposed to declarative construction. So I feel that it's much more natural for them to be after the instance read constructed.

You mention being able to do this, but I actively feel like that's not a good thing. 

Furthermore, you've claimed this is needed for dart/flutter compete. Can you point out how they allow for this functionality?  From my reading they seem to have the same limitation and would not allow arbitrary mutation during construction.

**Reply by najak3d** [reply]
My new proposal isn't mentioning Flutter.  I'm willing to concede that point.   They do not allow method calls from within their Init blocks.

**Reply by najak3d** [reply]
You wrote: "Methods imply state mutation as opposed to declarative construction. "

Yet methods are called from inside constructors often and without restraint, no?   So allowing the user to insert method calls into the Init-Block would be the same as this.

**Comment by najak3d** [comment]
What I'm fishing for right now- are the concerns regarding "allowing methods to be called prior to construction completion" that would allow a someone using a class to misuse that class by calling methods on it BEFORE construction is complete.   If there are real examples of how this might cause issues -- then I'm probably done here, and will retract the proposal, and not bring it up again.

**Comment by najak3d** [comment]
I think I might be ready to officially nix my own proposal.   I've learned a lot since I first made this proposal, and think I see sufficient reason now for rejecting it.

The solution we have now, "Fluent Pattern" is sufficient.   And modifying C# language to enable an equivalent notation without use of Extensions would be nice, but not mandatory.   The proposal I made has issues, and I don't see a viable path around these issues.  If someone else wants to propose a modification that makes this proposal viable -- then great.  I'll support that.

And also, it should be noted that the planned "Roles/Extensions" upgrade to the C# language (if it gets approved) will partially alleviate my concerns.   So if that gets done, then there is even less reason for my proposal.

So unless someone can convince me that this idea is worth pursuing -- I'm ready to concede, leaving behind nobody who thinks it's worth pursuing.   I'm OK with that.   I trust the overall judgement of the LDM.

**Comment by jmarolf** [comment]
In my opinion this proposal does both too much and too little.

#### Too Much

Implicitly ignoring return values seems dangerous. How do I know if the side effects are ok to ignore? Consider code like this:

```csharp
public interface ISettingsService : IDisposable
{
    string CollectionName { get; init; }
    bool IsRoamingEnabled { get; init; }
    bool HasPermissions(string scope);
}

public class SettingsService : ISettingsService
{
    // impl elided
}
```

Today I might initialize this code like so:

```csharp
var settingsService = new SettingsService()
{
    CollectionName = collectionName,
    IsRoamingEnabled = false
};

_ = settingsService.HasPermissions(Scopes.Default);
```

The discard (`_`) makes it clear to the reader that I am intentionally ignoring this value. I am hoping to add more warnings and analyzers to the .NET SDK so that people are made aware of places where they are ignoring return values. This is a general code smell and if I see code like this:

```csharp
settingsService.HasPermissions(Scopes.Default);
```

Today I am going to ask myself "Did they forget to do something with that value or is it fine to ignore?"

If we use your feature (as I understand it) we get something like this:

```csharp
var settingsService = new SettingsService()
{
    CollectionName = collectionName,
    IsRoamingEnabled = false
    .HasPermissions(Scopes.Default)
};
```

I now lack any ability to show my intent that the return value of `HasPermissions` is being intentionally ignored and is not an oversight.

There is also the issue that ignoring the return values may lead to hard to diagnose code defects unintentionally.

Take a type like this:

```csharp
public interface ISettingsService : IAsyncDisposable
{
    Task Init(string collectionName, bool isRoamingEnabled);
}

public class SettingsService : ISettingsService
{
    // implementation elided
}
```

Which would need to be initialized like this today:

```csharp
var settingsService = new SettingsService();
await settingsService.Init(collectionName, false);
```

If I were to use your feature it would look like this.

```csharp
var settingsService = new SettingsService()
{
    .Init(collectionName, false, 42)
};
```

Someone may unintentionally move `Init` into the object initializer as a simple refactoring without realizing that the `SettingsService` object is not properly initialized until `Init`'s return is awaited.

In general, I would be more onboard with this proposal if it followed similar semantic to https://github.com/dotnet/csharplang/discussions/781. If we restrict this feature to only allow `void` returning methods to be called in the object initializer we at least side-step this problem. I would consider this restriction a good step as there are many types in the BCL that would benefit still such as `List<T>`.

So this:

```csharp
public static List<string> RemoteFirstItem(List<string> strings){
    var list = new List<string>(strings);
    list.RemoveAt(0);
    return list
}
```

could become this:

```csharp
public static List<string> RemoteFirstItem(List<string> strings)
    => new List<string>(strings)
    {
        .RemoveAt(0)
    };
```

But we don't have the additional wrinkly of _encouraging_ developers to ignore return values from expressions.

#### Too Little

One of the main areas where the fluent builder pattern falls down today is async construction.

this works fine today:

```csharp
var app = builder.UseStaticFiles()
                 .UseRouting()
                 .UseAuthorization()
                 .MapRazorPages()
                 .Build();
```

but this is cumbersome:

```csharp
var app = (await (await (await (await builder.UseStaticFilesAsync())
         .UseRoutingAsync())
         .UseAuthorizationAsync())
         .MapRazorPagesAsync())
         .Build();
```

There are also no async constructors ([yet](https://github.com/dotnet/csharplang/discussions/419) :)) either so type like `SettingsService { Task Init(string collectionName, bool isRoamingEnabled); }` _must_ use a method call of some kind to initialize itself, it has no other choice. A proposal that attempts to make this easier would certainly add more value imho than what we have with this proposal right now. In its current state, this proposal syntactic sugar that does not allow you do anything new nor provide additional type-safety or runtime-safety to the developer.

As you've hit on a bit of the problem in the discussion here. Complex object construction can be difficult today and there are existing problems about ensuring things are definitely-assigned, correctly non-null, and fully initialized. Perhaps there is a much bigger proposal about having some kind of "initialization scope" that operates very differently from the object initializer block and allows for the compiler and developer to work hand in hand to ensure complex objects are constructed correctly.

tl;dr this proposal needs to drastically narrow in scope so that its cost does not outstrip its value or greatly increase its scope so that it attempts to tackle more fundamental problems in object construction. As it currently stands in the middle, I don't personally think I would take it.

I hope my words aren't discouraging @najak3d. If you have any more insights into the problems you are having, I would love to hear about it. The whole point of this repo is to talk these things through and see how well they stand up to scrutiny.

**Reply by najak3d** [reply]
@jmarolf - I would vote to narrow the scope, to only allowing "voids" to be called, and no 'async' methods.   If async is involved, you've got to find another way.    Init-Blocks are non-async already -- and we can keep it that way.   Object construction is typically a fast single-threaded synchronous operation anyways.

So making it "less" would be my preference.

The big issue that remains is that allowing any void method to be called by the init-block would make it troublesome for some classes.  Would they now have to test every method call from within the context of being run before construction is complete?

If this proposal got pushed through, IMO, at minimum we'd need to invent a new Method-Attribute called "[InitCallable(true/false)]" that permits a class designed to NOT allow void methods to be called.  Or it could be a class-attribute for the mode of that class's permissiveness on calling methods from within the Init-block: "

"Disabled" - no class methods can be called in Init Block for that class
"OptIn" -  only class methods with "[InitCallable(true)]" can be called.
"OptOut" - only class methods with "[InitCallable(false)]" can NOT be called  

The use of these attributes would enable class designers to turn this feature off (or limits it's use) at their whim.

By default, I'd bet 90%+ of the situations would be fine with calling any of their "void" methods from within the Init block.

Also, I suppose the entire DLL/assembly can have an attribute on it, so that you can turn this off/on for the entire assembly, and set it to "Disable, OptIn, or OptOut" for all classes in that assembly.

Additionally, if we are afraid of risks, it could be something that is not backward compatible -- maybe you can only use this notation for assemblies compiled with C#13+ that support this usage, thus giving the developers a fair chance to disable it before release.   I'm not sure how this works out logistically, if this is a reasonable safeguard or not.

**Reply by najak3d** [reply]
One could say "limiting to voids" is too restrictive, but if there are instances where non-voids really need to be called, those specific methods can each be remedied via Extension Method, which does the same thing as the targeted method, but returns a void.  This would be a sparse use of Extensions, targeted only at a few methods.

As is now, to get a rich initialization capability with the Init-Block, it needs to support method calls -- or you have to create Fluent-style Extensions for every member that you want to set (including properties) - and so it becomes an All-or-Nothing approach, since you can't interchange methods with property setters -- currently.

If it is deemed "not too much risk", I would still like to see ".method()" calls that return void, permitted.

I'm going to modify my original post, to reflect the "void" requirement.

**Reply by najak3d** [reply]
I revamped the proposal to:
1. Mention "void" only methods
2. And to suggest combining with #3578, as an improvement on that proposal by tightening it's scope for where these methods could be called.  (easier for a compiler to recognize the init-block syntax, vs outside of the init-block)

**Comment by najak3d** [comment]
I'm now trying to give this one-more-shot, by incorporating feedback from @jmarolf , and by tying it into existing proposals here:

[https://github.com/dotnet/csharplang/issues/5176](https://github.com/dotnet/csharplang/issues/5176)

and

[https://github.com/dotnet/csharplang/discussions/3578](https://github.com/dotnet/csharplang/discussions/3578)

===
**The proposal I am trying to make would:**
1. Fully accommodate the objectives of these other proposals, as well as 
2. Enables "init-blocks" to mostly eliminate the need for "Fluent-Pattern" extensions to provide a similar simple declarative syntax.

Classes, all by themselves will be able to enable a terse Fluent-like syntax for object initialization, without creating oddball methods that "return the instance" (a requirement for fluent-syntax today).  It would "just work", out-of-box.

At same time, it would allow for full-initialization of objects that make use of "readonly" specifiers for many values, which might need to be set from inside an "init method".

All this, while only requesting a simple compiler upgrade, to simply interpret the ".method()" notation similar to a "wither" block, and employ very simple rules on "which methods can be called", and a tight scope, since it all happens within the Init-block directly.

Risk would be minimized by defaulting this feature to off, and only methods that the class designer specifies will be callable from within the init-block.  Thus we mitigate the risk of class-users doing things that the class designer didn't intend.  It all remains locked-down.

**Comment by najak3d** [comment]
Looking again here - I think my new spec might be "overly locked down". 

If the non-init ".method()" are called on a fully instantiated object - there is truly NO RISK to allowing ANY "void" method to be called AFTER the Init-Properties/Methods are all called.

For example, there's no reason that a class-designer should EVER care if you call other methods that can be called on an instantiated object.  There is no risk here.  No reason for the "lock-down".

Suggested Pattern should be:

```c#
new MyType()
{
    InitProperty = value,
   .InitMethodCall1(args),
   .InitMethodCall2(args),   
   .NonInitProperty = value2, // <<== **Object becomes fully instantiated after this line of code runs**
   .ExtensionProperty = value3,  // Only applies if the new "Roles/Extensions" proposal is implemented
   .NonInitMethodCall1(args),  // COULD be an Extension
   .GridPos(2 ,3),
   .Children {  add children here  }
}
```
In short, "Construction completes" (where readonly fields become immutable) after the final Property or Init-Method is called.  And so any calls made to methods after that will not be able to modify "readonly" fields.    Only methods prefixed with "Init" will be able to modify these readonly-fields.

Thus we do NOT need the "[CallableInitMethod]" attributes at all, because all "void methods" can now be called from inside the Init-block, and Init-Methods can ALL be called from within the Init-Block, but simply must precede ALL non-init method calls.

**Comment by najak3d** [comment]
OK - I revamped the Opening statement once more -- simplifying it, removing the suggestion of requiring Attributes to enable the calling of "void" methods, since these pose no risk to the class-designer, because the non-init method calls will be operating the same as if they were called using the long-hand notation after the Init Block.

I believe it now accommodates the objectives that I originally had in mind, and also the objectives of two existing proposals:  #3578 and #5176.  And also now accounts for the feedback from @jmarolf  (above).

Can this proposal now grow some legs?

**Comment by najak3d** [comment]
Regarding concerns about "calling methods from inside an Init-block, and the "dangers" of this.  

As I've considered this more -- if the class is NOT SEALED, then you can *already* create a new class derivation from that class, and inside of its constructor call ANY METHOD from inside the constructor BEFORE it is fully initialized.   So these dangers already exist.

So this danger *already* exists, and so allowing method calls from within the Init-Block introduce no new dangers, and the notation is very intuitive.
